gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspConfigurations.kt

```diff
@@ -86,6 +86,16 @@ class KspConfigurations(private val project: Project) {
             // 1.6.0: decorateKotlinProject(project.kotlinExtension)?
             decorateKotlinProject(project.extensions.getByName("kotlin") as KotlinProjectExtension, project)
         }
+        // Create sourceSet-specific KSP configurations for the case when the KotlinBaseApiPlugin is applied instead
+        // of the KotlinBasePluginWrapper (e.g., when AGP's built-in Kotlin support is enabled).
+        project.plugins.withType(KotlinBaseApiPlugin::class.java) {
+            AndroidPluginIntegration.forEachAndroidSourceSet(project) { sourceSet ->
+                createConfiguration(
+                    name = getAndroidConfigurationName(kotlinTargetName = "", sourceSet),
+                    readableSetName = "$sourceSet (Android)"
+                )
+            }
+        }
```

Comment:
It looks like the `kotlin` extension [can be obtained](https://github.com/JetBrains/kotlin/blob/master/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinBaseApiPlugin.kt#L64) from `KotlinBaseApiPlugin`. Have you tried to just call `decorateKotlinProject(...)`?
1. It will cover the case where `KotlinBaseApiPlugin` is applied in the context other than AGP.
2. also much easier to maintain.

If `kotlin` doesn't work, either it should be initialized properly by AGP, and/or the configurations should only be created when AGP's built-in Kotlin compilation is applied. From maintenance's point of view, I'd also hoist the common code of `AndroidPluginIntegration.forEachAndroidSourceSet...` into a private function.

Reply:
> It looks like the kotlin extension can be obtained from KotlinBaseApiPlugin. Have you tried to just call decorateKotlinProject(...)?

I think that's the right long-term approach, but currently AGP doesn't add the `kotlin` extension (block by [KT-70897](https://youtrack.jetbrains.com/issue/KT-70897/Add-KotlinBaseApiPlugin.kotlinAndroidExtension)). After that is fixed and AGP adds the `kotlin` extension, I can modify this to instead call `decorateKotlinProject()`.

> From maintenance's point of view, I'd also hoist the common code of AndroidPluginIntegration.forEachAndroidSourceSet... into a private function.

Done